File: packages/adapters/src/venues/pionex.ts
import { type FundingEvent, type FundingSnapshot, inferIntervalHours } from "@ai-rates/core";
import { CircuitOpenError } from "../http";
import { hoursBetween, marketRef, mul, num, selectRefreshBatch } from "../parse";
import type { VenueAdapter } from "../types";
import { basisHoursFromGaps } from "./aster";

const VENUE_ID = "pionex";
export const PIONEX_API = "https://api.pionex.com/api/v1";
/** The PERP symbol list is 250 KB and weight 5 of a 10-per-second budget, so hourly. */
export const SYMBOLS_MAX_AGE_MS = 60 * 60_000;
/**
 * Pionex publishes no settlement interval in any bulk response, and the interval varies (1h, 4h
 * and 8h all trade), so it is read per symbol from the spacing of its last settlements. At this
 * budget a cold start covers the ~570 dollar perps in about 19 cycles; `warmUp` skips that.
 */
export const INTERVAL_REFRESH_BUDGET = 30;
export const INTERVAL_MAX_AGE_MS = 6 * 60 * 60_000;
/** A symbol with no settlement yet has no interval to read; look again after this long. */
const INTERVAL_RETRY_MS = 30 * 60_000;
const INTERVAL_SAMPLE = 4;
/** `limit` above 100 is refused with MARKET_PARAMETER_ERROR (measured 2026-09-14). */
const HISTORY_PAGE_SIZE = 100;
const HISTORY_MAX_PAGES = 50;
/**
 * Quotes collected. 43 of Pionex's 610 perps on 2026-09-14 price one coin in another (BTC_ETH_PERP,
 * PAXG_BTC_PERP, QQQX_SPYX_PERP), so their open interest and volume are not dollars and their mark
 * belongs in no USD pool.
 */
const DOLLAR_QUOTES: ReadonlySet<string> = new Set(["USDT", "USDC"]);

export interface PionexEnvelope<T> {
  result: boolean;
  data: T;
  code?: string;
  message?: string;
  timestamp?: number;
}

export interface PionexSymbol {
  symbol: string;
  type: string;
  /** Pionex's own coin code, which is not always the ticker: ZEROG for 0G, LIGHTER for LIT. */
  baseCurrency: string;
  quoteCurrency: string;
  status: string;
}

export interface PionexIndex {
  symbol: string;
  indexPrice: string;
  markPrice: string;
  /** Rate to be paid at `nextFundingTime`, per settlement interval. */
  nextFundingRate: string;
  nextFundingTime: number;
}

export interface PionexTicker {
  symbol: string;
  /** 24h volume in base units. */
  volume: string;
  /** 24h turnover in the quote currency. */
  amount: string;
}

export interface PionexOpenInterest {
  symbol: string;
  /** Base units. */
  openInterest: string;
}

export interface PionexBookTicker {
  symbol: string;
  bidPrice: string;
  /** Base units. */
  bidSize: string;
  askPrice: string;
  askSize: string;
}

export interface PionexFundingRate {
  fundingRate: string;
  fundingTime: number;
}

export interface PionexIntervalEntry {
  hours: number | null;
  fetchedAt: number;
}

function unwrap<T>(json: PionexEnvelope<T>, what: string): T {
  if (!json?.result || json.data === undefined || json.data === null) {
    throw new Error(`pionex ${what}: ${json?.code ?? ""} ${json?.message ?? ""}`);
  }
  return json.data;
}

/** Dollar-quoted perps Pionex lists as TRADING, by symbol. */
export function tradablePionexPerps(symbols: readonly PionexSymbol[]): Map<string, PionexSymbol> {
  return new Map(
    symbols
      .filter(
        (s) => s.type === "PERP" && s.status === "TRADING" && DOLLAR_QUOTES.has(s.quoteCurrency),
      )
      .map((s) => [s.symbol, s]),
  );
}

/**
 * Settlement interval from a symbol's most recent settlements (newest first, as served).
 *
 * The median gap when there are two or more: ACT_USDT_PERP settled every 4h, ACH_USDT_PERP every 8h
 * and AAX_USDT_PERP every hour on 2026-09-14. A listing with a single settlement so far is measured
 * against its next one instead.
 */
export function pionexIntervalHours(
  rates: readonly PionexFundingRate[],
  nextFundingTime: number | null,
): number | null {
  const times = rates.map((r) => num(r.fundingTime)).filter((t): t is number => t !== null);
  if (times.length >= 2) return inferIntervalHours(times);
  if (times.length === 1) return hoursBetween(times[0] as number, nextFundingTime);
  return null;
}

export interface PionexSnapshotInput {
  symbols: ReadonlyMap<string, PionexSymbol>;
  indexes: readonly PionexIndex[];
  tickers: readonly PionexTicker[];
  openInterests: readonly PionexOpenInterest[];
  bookTickers: readonly PionexBookTicker[];
  intervals: ReadonlyMap<string, PionexIntervalEntry>;
}

/**
 * Joins the bulk index, ticker, open-interest and book responses onto tradable perps whose interval
 * is known. A perp appears once its interval has been read.
 *
 * Units, checked against the live responses of 2026-09-14 22:01 UTC:
 * - `nextFundingRate` is a fraction per settlement interval: the resting rate is 0.0001 on 8h perps
 *   (ACH), 0.00005 on 4h perps (ACT), and each of those settled at exactly that in the history.
 * - `openInterest` is base units: BTC_USDT_PERP 1,335.1442 x mark 77,096 = $102.9M, against 24h
 *   turnover of $1.99bn. Read as dollars it would be $1,335 of BTC open interest.
 * - `amount` is quote turnover: BTC volume 25,822.66 x the day's range 76,463-77,423 brackets
 *   1,989,264,906 (average 77,036).
 * - Book sizes are base units, like order sizes (`minSizeLimit` 0.0001 BTC).
 *
 * Pionex declares no asset class anywhere in its public API, so every perp is crypto. Its AAPLX,
 * TSLAX and friends are xStocks tokens in any case.
 */
export function parsePionexSnapshots(input: PionexSnapshotInput, now: number): FundingSnapshot[] {
  const tickers = new Map(input.tickers.map((t) => [t.symbol, t]));
  const openInterest = new Map(input.openInterests.map((o) => [o.symbol, num(o.openInterest)]));
  const books = new Map(input.bookTickers.map((b) => [b.symbol, b]));
  const snapshots: FundingSnapshot[] = [];

  for (const index of input.indexes) {
    const symbol = input.symbols.get(index.symbol);
    const rate = num(index.nextFundingRate);
    const hours = input.intervals.get(index.symbol)?.hours ?? null;
    if (!symbol || rate === null || hours === null || hours <= 0) continue;

    const markPrice = num(index.markPrice);
    const next = num(index.nextFundingTime);
    const book = books.get(index.symbol);
    const bestBid = num(book?.bidPrice);
    const bestAsk = num(book?.askPrice);
    snapshots.push({
      ...marketRef(VENUE_ID, index.symbol, { quote: symbol.quoteCurrency }),
      observedAt: now,
      rate,
      basisHours: hours,
      intervalHours: hours,
      nextFundingAt: next !== null && next > 0 ? next : null,
      kind: "predicted",
      markPrice,
      indexPrice: num(index.indexPrice),
      bestBid,
      bestBidSizeUsd: mul(num(book?.bidSize), bestBid),
      bestAsk,
      bestAskSizeUsd: mul(num(book?.askSize), bestAsk),
      openInterestUsd: mul(openInterest.get(index.symbol) ?? null, markPrice),
      volume24hUsd: num(tickers.get(index.symbol)?.amount),
    });
  }
  return snapshots;
}

/**
 * Settlements for one perp in [fromMs, toMs], oldest first. Basis hours come from the gaps between
 * every fetched settlement, including the neighbours just outside the window.
 */
export function parsePionexFundingHistory(
  venueSymbol: string,
  rows: readonly PionexFundingRate[],
  fromMs: number,
  toMs: number,
  fallbackHours: number | null,
  quote?: string,
): FundingEvent[] {
  const byTime = new Map<number, number>();
  for (const row of rows) {
    const time = num(row.fundingTime);
    const rate = num(row.fundingRate);
    if (time !== null && rate !== null) byTime.set(time, rate);
  }
  const times = [...byTime.keys()].sort((a, b) => a - b);
  const basis = basisHoursFromGaps(times, fallbackHours);
  const ref = marketRef(VENUE_ID, venueSymbol, quote ? { quote } : {});

  const events: FundingEvent[] = [];
  times.forEach((settledAt, i) => {
    const basisHours = basis[i];
    if (settledAt < fromMs || settledAt > toMs || basisHours === null || basisHours === undefined) {
      return;
    }
    events.push({
      ...ref,
      settledAt,
      rate: byTime.get(settledAt) as number,
      basisHours,
      markPrice: null,
    });
  });
  return events;
}

export interface PionexAdapterOptions {
  intervalRefreshBudget?: number;
}

export function createPionexAdapter(options: PionexAdapterOptions = {}): VenueAdapter {
  const budget = options.intervalRefreshBudget ?? INTERVAL_REFRESH_BUDGET;
  let symbols: { fetchedAt: number; bySymbol: Map<string, PionexSymbol> } | null = null;
  const intervals = new Map<string, PionexIntervalEntry>();

  const fundingRatesUrl = (symbol: string, limit: number, endTime?: number) =>
    `${PIONEX_API}/market/fundingRates?symbol=${encodeURIComponent(symbol)}${endTime === undefined ? "" : `&endTime=${endTime}`}&limit=${limit}`;

  return {
    venueId: VENUE_ID,
    // 10 weight per second per IP, shared by every endpoint; the symbol list alone weighs 5.
    minIntervalMs: 150,

    /** The stored interval is what the history would say again; fetchedAt 0 keeps it due for a re-read. */
    warmUp(markets) {
      for (const market of markets) {
        const hours = market.intervalHours;
        if (hours !== null && hours > 0 && !intervals.has(market.venueSymbol)) {
          intervals.set(market.venueSymbol, { hours, fetchedAt: 0 });
        }
      }
    },

    async fetchSnapshots(client, now) {
      if (!symbols || now - symbols.fetchedAt >= SYMBOLS_MAX_AGE_MS) {
        const data = unwrap(
          await client.getJson<PionexEnvelope<{ symbols: PionexSymbol[] }>>(
            `${PIONEX_API}/common/symbols?type=PERP`,
          ),
          "symbols",
        );
        symbols = { fetchedAt: now, bySymbol: tradablePionexPerps(data.symbols ?? []) };
      }
      const live = symbols.bySymbol;

      const [indexes, tickers, openInterests, bookTickers] = await Promise.all([
        client
          .getJson<PionexEnvelope<{ indexes: PionexIndex[] }>>(`${PIONEX_API}/market/indexes`)
          .then((json) => unwrap(json, "indexes").indexes ?? []),
        client
          .getJson<PionexEnvelope<{ tickers: PionexTicker[] }>>(
            `${PIONEX_API}/market/tickers?type=PERP`,
          )
          .then((json) => unwrap(json, "tickers").tickers ?? []),
        client
          .getJson<PionexEnvelope<{ openInterests: PionexOpenInterest[] }>>(
            `${PIONEX_API}/market/openInterests?type=PERP`,
          )
          .then((json) => unwrap(json, "open interests").openInterests ?? []),
        client
          .getJson<PionexEnvelope<{ tickers: PionexBookTicker[] }>>(
            `${PIONEX_API}/market/bookTickers?type=PERP`,
          )
          .then((json) => unwrap(json, "book tickers").tickers ?? []),
      ]);

      for (const symbol of intervals.keys()) {
        if (!live.has(symbol)) intervals.delete(symbol);
      }
      const nextFunding = new Map(indexes.map((i) => [i.symbol, num(i.nextFundingTime)]));
      const liveSymbols = indexes.map((i) => i.symbol).filter((s) => live.has(s));
      for (const symbol of selectRefreshBatch(
        liveSymbols,
        intervals,
        now,
        budget,
        INTERVAL_MAX_AGE_MS,
      )) {
        try {
          const data = unwrap(
            await client.getJson<PionexEnvelope<{ rates: PionexFundingRate[] }>>(
              fundingRatesUrl(symbol, INTERVAL_SAMPLE),
            ),
            `funding rates ${symbol}`,
          );
          const hours = pionexIntervalHours(data.rates ?? [], nextFunding.get(symbol) ?? null);
          if (hours !== null && hours > 0) {
            intervals.set(symbol, { hours, fetchedAt: now });
          } else {
            // Nothing settled yet. Keep any interval already known, and come back sooner than the
            // full max age by back-dating the entry.
            const known = intervals.get(symbol)?.hours ?? null;
            intervals.set(symbol, {
              hours: known,
              fetchedAt: now - INTERVAL_MAX_AGE_MS + INTERVAL_RETRY_MS,
            });
          }
        } catch (error) {
          if (error instanceof CircuitOpenError) break;
          // Leave this symbol for a later cycle; one bad symbol shouldn't fail the batch.
        }
      }

      const snapshots = parsePionexSnapshots(
        { symbols: live, indexes, tickers, openInterests, bookTickers, intervals },
        now,
      );
      return { snapshots, settled: [] };
    },

    async fetchFundingHistory(client, venueSymbol, fromMs, toMs) {
      // Newest first; `endTime` is inclusive and `startTime` is ignored, so page backwards from toMs.
      const rows: PionexFundingRate[] = [];
      let endTime = toMs;
      for (let page = 0; page < HISTORY_MAX_PAGES && endTime >= fromMs; page++) {
        const list =
          unwrap(
            await client.getJson<PionexEnvelope<{ rates: PionexFundingRate[] }>>(
              fundingRatesUrl(venueSymbol, HISTORY_PAGE_SIZE, endTime),
            ),
            "funding rates",
          ).rates ?? [];
        rows.push(...list);
        if (list.length < HISTORY_PAGE_SIZE) break;
        endTime = Math.min(...list.map((r) => Number(r.fundingTime))) - 1;
      }
      return parsePionexFundingHistory(
        venueSymbol,
        rows,
        fromMs,
        toMs,
        intervals.get(venueSymbol)?.hours ?? null,
        symbols?.bySymbol.get(venueSymbol)?.quoteCurrency,
      );
    },
  };
}

export const pionexAdapter: VenueAdapter = createPionexAdapter();

File: apps/collector-go/internal/adapters/pionex/pionex.go
// Package pionex parses Pionex's perpetual futures.
//
// Ported from packages/adapters/src/venues/pionex.ts. Two venue quirks shape every function here:
//
//   - PIONEX PUBLISHES NO SETTLEMENT INTERVAL in any bulk response, and the interval varies (1h, 4h
//     and 8h all trade), so it is read per symbol from the spacing of its last settlements. A perp
//     appears only once its interval is known, which is why the adapter's cache -- and WarmUp, which
//     seeds it -- is load-bearing rather than an optimisation.
//   - SIZES AND OPEN INTEREST ARE BASE UNITS, not dollars and not contracts. BTC_USDT_PERP's
//     openInterest of 1,335.1442 x mark 77,096 is $102.9M, against 24h turnover of $1.99bn; read as
//     dollars it would be $1,335 of open interest on a market doing two billion a day.
package pionex

import (
	"fmt"
	"sort"

	"github.com/belyjelli/ai-rates/collector/internal/adapters"
	"github.com/belyjelli/ai-rates/collector/internal/core"
)

const VenueID = "pionex"

// API is the versioned REST root every endpoint hangs off.
const API = "https://api.pionex.com/api/v1"

const (
	// SymbolsMaxAgeMs: the PERP symbol list is 250 KB and weight 5 of a 10-per-second budget, so
	// hourly.
	SymbolsMaxAgeMs int64 = 60 * 60_000

	// IntervalRefreshBudget is per-symbol fundingRates calls allowed per cycle. At this budget a cold
	// start covers the ~570 dollar perps in about 19 cycles; WarmUp skips that.
	IntervalRefreshBudget = 30
	// IntervalMaxAgeMs: re-read each symbol's interval this often.
	IntervalMaxAgeMs int64 = 6 * 60 * 60_000
	// intervalRetryMs: a symbol with no settlement yet has no interval to read; look again after this
	// long, which is done by back-dating its cache entry rather than by a second timestamp.
	intervalRetryMs int64 = 30 * 60_000
	// intervalSample is how many settlements are enough to read the spacing from.
	intervalSample = 4

	// historyPageSize: `limit` above 100 is refused with MARKET_PARAMETER_ERROR (measured
	// 2026-09-14).
	historyPageSize = 100
	historyMaxPages = 50
)

// dollarQuotes are the quotes collected. 43 of Pionex's 610 perps on 2026-09-14 price one coin in
// another (BTC_ETH_PERP, PAXG_BTC_PERP, QQQX_SPYX_PERP), so their open interest and volume are not
// dollars and their mark belongs in no USD pool.
var dollarQuotes = map[string]struct{}{"USDT": {}, "USDC": {}}

// Envelope is Pionex's uniform response wrapper.
//
// Data is a pointer so an absent or null `data` is distinguishable from an empty one, which is what
// the TypeScript `data === undefined || data === null` guard turns into a thrown error: a response
// that lost its payload must fail the cycle rather than read as a venue with nothing listed.
type Envelope[T any] struct {
	Result    bool   `json:"result"`
	Data      *T     `json:"data"`
	Code      string `json:"code"`
	Message   string `json:"message"`
	Timestamp int64  `json:"timestamp"`
}

func (e Envelope[T]) unwrap(what string) (T, error) {
	if !e.Result || e.Data == nil {
		var zero T
		return zero, fmt.Errorf("pionex %s: %s %s", what, e.Code, e.Message)
	}
	return *e.Data, nil
}

// The payload shapes, one per endpoint. Pionex nests every list under its own key, and the book
// endpoint reuses `tickers` for a different row shape than the ticker endpoint does.
type (
	SymbolList struct {
		Symbols []Symbol `json:"symbols"`
	}
	IndexList struct {
		Indexes []Index `json:"indexes"`
	}
	TickerList struct {
		Tickers []Ticker `json:"tickers"`
	}
	OpenInterestList struct {
		OpenInterests []OpenInterest `json:"openInterests"`
	}
	BookTickerList struct {
		Tickers []BookTicker `json:"tickers"`
	}
	FundingRateList struct {
		Symbol string        `json:"symbol"`
		Rates  []FundingRate `json:"rates"`
	}
)

type Symbol struct {
	Symbol string `json:"symbol"`
	Type   string `json:"type"`
	// BaseCurrency is Pionex's own coin code, which is not always the ticker: ZEROG for 0G, LIGHTER
	// for LIT. The symbol's own base is preferred, so this is read only for the quote's sake.
	BaseCurrency  string `json:"baseCurrency"`
	QuoteCurrency string `json:"quoteCurrency"`
	Status        string `json:"status"`
}

type Index struct {
	Symbol     string       `json:"symbol"`
	IndexPrice adapters.Num `json:"indexPrice"`
	MarkPrice  adapters.Num `json:"markPrice"`
	// NextFundingRate is the rate to be paid at NextFundingTime, per settlement interval.
	NextFundingRate adapters.Num `json:"nextFundingRate"`
	NextFundingTime adapters.Num `json:"nextFundingTime"`
}

type Ticker struct {
	Symbol string `json:"symbol"`
	// Volume is 24h volume in BASE units.
	Volume adapters.Num `json:"volume"`
	// Amount is 24h turnover in the QUOTE currency.
	Amount adapters.Num `json:"amount"`
}

type OpenInterest struct {
	Symbol string `json:"symbol"`
	// OpenInterest is BASE units.
	OpenInterest adapters.Num `json:"openInterest"`
}

type BookTicker struct {
	Symbol   string       `json:"symbol"`
	BidPrice adapters.Num `json:"bidPrice"`
	// BidSize and AskSize are BASE units, like order sizes (`minSizeLimit` 0.0001 BTC).
	BidSize  adapters.Num `json:"bidSize"`
	AskPrice adapters.Num `json:"askPrice"`
	AskSize  adapters.Num `json:"askSize"`
}

type FundingRate struct {
	FundingRate adapters.Num `json:"fundingRate"`
	FundingTime adapters.Num `json:"fundingTime"`
}

// IntervalEntry is one symbol's cached settlement interval.
//
// Hours is a pointer because a listing that has not settled yet has no interval to cache, and that
// is different from having no entry at all: an entry with nil hours is what holds the retry
// back-dating, so the symbol is re-read in 30 minutes rather than immediately every cycle.
type IntervalEntry struct {
	Hours     *float64
	FetchedAt int64
}

// TradablePerps is the dollar-quoted perps Pionex lists as TRADING, by symbol.
func TradablePerps(symbols []Symbol) map[string]Symbol {
	live := make(map[string]Symbol, len(symbols))
	for _, symbol := range symbols {
		if symbol.Type != "PERP" || symbol.Status != "TRADING" {
			continue
		}
		if _, dollar := dollarQuotes[symbol.QuoteCurrency]; !dollar {
			continue
		}
		live[symbol.Symbol] = symbol
	}
	return live
}

// IntervalHours is the settlement interval from a symbol's most recent settlements (newest first, as
// served).
//
// The median gap when there are two or more: ACT_USDT_PERP settled every 4h, ACH_USDT_PERP every 8h
// and AAX_USDT_PERP every hour on 2026-09-14. A listing with a single settlement so far is measured
// against its next one instead.
func IntervalHours(rates []FundingRate, nextFundingTime *int64) *float64 {
	times := make([]int64, 0, len(rates))
	for _, rate := range rates {
		if rate.FundingTime.OK {
			times = append(times, int64(rate.FundingTime.Val))
		}
	}
	if len(times) >= 2 {
		return core.InferIntervalHours(times)
	}
	if len(times) == 1 {
		return adapters.HoursBetween(&times[0], nextFundingTime)
	}
	return nil
}

// SnapshotInput is one cycle's bulk responses plus the interval cache they are joined against.
type SnapshotInput struct {
	Symbols       map[string]Symbol
	Indexes       []Index
	Tickers       []Ticker
	OpenInterests []OpenInterest
	BookTickers   []BookTicker
	Intervals     map[string]IntervalEntry
}

// ParseSnapshots joins the bulk index, ticker, open-interest and book responses onto tradable perps
// whose interval is known. A perp appears once its interval has been read.
//
// Units, checked against the live responses of 2026-09-14 22:01 UTC:
//   - `nextFundingRate` is a fraction per settlement interval: the resting rate is 0.0001 on 8h perps
//     (ACH), 0.00005 on 4h perps (ACT), and each of those settled at exactly that in the history.
//   - `openInterest` is base units: BTC_USDT_PERP 1,335.1442 x mark 77,096 = $102.9M, against 24h
//     turnover of $1.99bn. Read as dollars it would be $1,335 of BTC open interest.
//   - `amount` is quote turnover: BTC volume 25,822.66 x the day's range 76,463-77,423 brackets
//     1,989,264,906 (average 77,036).
//   - Book sizes are base units, like order sizes (`minSizeLimit` 0.0001 BTC).
//
// Pionex declares no asset class anywhere in its public API, so every perp is crypto. Its AAPLX,
// TSLAX and friends are xStocks tokens in any case.
//
// The output follows the index response's own order, so it does not reshuffle per run the way
// ranging a Go map would.
func ParseSnapshots(input SnapshotInput, now int64) []core.FundingSnapshot {
	tickers := make(map[string]Ticker, len(input.Tickers))
	for _, ticker := range input.Tickers {
		tickers[ticker.Symbol] = ticker
	}
	openInterest := make(map[string]adapters.Num, len(input.OpenInterests))
	for _, entry := range input.OpenInterests {
		openInterest[entry.Symbol] = entry.OpenInterest
	}
	books := make(map[string]BookTicker, len(input.BookTickers))
	for _, book := range input.BookTickers {
		books[book.Symbol] = book
	}

	snapshots := make([]core.FundingSnapshot, 0, len(input.Indexes))
	for _, index := range input.Indexes {
		symbol, listed := input.Symbols[index.Symbol]
		hours := input.Intervals[index.Symbol].Hours
		if !listed || !index.NextFundingRate.OK || hours == nil || *hours <= 0 {
			continue
		}

		markPrice := index.MarkPrice.Ptr()
		book := books[index.Symbol]
		bestBid := book.BidPrice.Ptr()
		bestAsk := book.AskPrice.Ptr()
		// The quote is what the venue lists, not what the symbol implies: Pionex's coin-quoted perps
		// are excluded upstream precisely because that distinction is real money.
		quote := symbol.QuoteCurrency
		basisHours := *hours
		intervalHours := *hours

		snapshots = append(snapshots, core.FundingSnapshot{
			MarketRef:       adapters.MarketRefFor(VenueID, index.Symbol, adapters.Overrides{Quote: &quote, HasQuote: true}),
			ObservedAt:      now,
			Rate:            index.NextFundingRate.Val,
			BasisHours:      basisHours,
			IntervalHours:   &intervalHours,
			NextFundingAt:   index.NextFundingTime.PositiveMs(),
			Kind:            core.KindPredicted,
			MarkPrice:       markPrice,
			IndexPrice:      index.IndexPrice.Ptr(),
			BestBid:         bestBid,
			BestBidSizeUSD:  adapters.Mul(book.BidSize.Ptr(), bestBid),
			BestAsk:         bestAsk,
			BestAskSizeUSD:  adapters.Mul(book.AskSize.Ptr(), bestAsk),
			OpenInterestUSD: adapters.Mul(openInterest[index.Symbol].Ptr(), markPrice),
			Volume24hUSD:    tickers[index.Symbol].Amount.Ptr(),
		})
	}
	return snapshots
}

// ParseFundingHistory returns settlements for one perp in [fromMs, toMs], oldest first. Basis hours
// come from the gaps between every fetched settlement, including the neighbours just outside the
// window -- which is why the window is applied after the basis is computed, not before.
//
// quote is what the venue lists for the symbol, when the adapter still has its symbol cache; nil
// falls back to parsing the symbol.
func ParseFundingHistory(
	venueSymbol string,
	rows []FundingRate,
	fromMs, toMs int64,
	fallbackHours *float64,
	quote *string,
) []core.FundingEvent {
	byTime := make(map[int64]float64, len(rows))
	for _, row := range rows {
		if row.FundingTime.OK && row.FundingRate.OK {
			byTime[int64(row.FundingTime.Val)] = row.FundingRate.Val
		}
	}
	times := make([]int64, 0, len(byTime))
	for at := range byTime {
		times = append(times, at)
	}
	sort.Slice(times, func(i, j int) bool { return times[i] < times[j] })

	basis := adapters.BasisHoursFromGaps(times, fallbackHours)
	ref := adapters.MarketRefFor(VenueID, venueSymbol, adapters.Overrides{Quote: quote, HasQuote: quote != nil})

	events := make([]core.FundingEvent, 0, len(times))
	for i, settledAt := range times {
		basisHours := basis[i]
		if settledAt < fromMs || settledAt > toMs || basisHours == nil {
			continue
		}
		events = append(events, core.FundingEvent{
			MarketRef:  ref,
			SettledAt:  settledAt,
			Rate:       byTime[settledAt],
			BasisHours: *basisHours,
			MarkPrice:  nil,
		})
	}
	return events
}
